README.md

The purpose of the WID-qEC is to estimate the risk for endometrial and/or cervical cancer based on the methylation status of DNA recovered from cervical smear samples, assayed with MethyLight.

The CalculateWIDqEC App needs the exported "Results file" from Quantstudio 5 Design & Analysis Software in as tab-separated .txt, and assumes the export has been done using the correct sampleIDs, targets and fixed Ct thresholds.

## Instructions

Make sure R is installed and in your environment path.

Make sure your default browser is a modern one.

Libraries needed: shiny, dplyr, tidyverse, openxlsx, ggplot2, ggpubr.

Double-click CalculateWIDqEC.bat (or a short-cut thereof), and follow instructions in the browser.

CHANGELOG.md
## [0.4.1] - 31.01.2023

## Changed
* Removed ',' from QC explanation in final .csv file, because of troubles with import into Swisslab

## [0.4.0] - 19.01.2023

## Changed
* CT threshold all targets excpet COL2A1 now 35 (more stringent, previously 38)
* if GYPC1, GYPC2 or ZSCAN12 ampilfies only in one of two replicates, this amplification is disregarded. 
* lowered threshold for warning high CT to 30 (previously at 35).

## [0.2.0] - 03.11.2022

## Added
* output additional list checking whether SD for COL2A1 between reps > 1.5 CT, gives warning in QC for WID-qEC
* plot with mean+SD CT COL2A1 for all samples on plate

## [0.1.0] - 31.10.2022

### Changed
* CT threshold COL2A1 now 30 (more stringent, previously 35)

### Added
* Additional CT threshold for other targets of 38


## [0.0.1] - 01.09.2022

Initial established version


src/App.R

# input is fixed to uploading one results file exported from Quantstudio 5, .txt format
# ! will have to check naming of the standards and controls!!!
# output files are in a .zip
# pic of the standard curve
# Excel workbook with all processed info (PMR for all targets, Ct+SD, DNA input, analysis reps)
# Final csv file that is imported for commercial WIDqEC into Swisslab

# load libraries
library(shiny)
library(dplyr)
library(tidyverse)
library(openxlsx)
library(ggplot2)
library(ggpubr)

# source main calculate_PMR function
source("./calculate_pmr_widqec.R")

# Define UI ----
ui <- fluidPage(
  # App title ----
  titlePanel(h3("Calculate PMR and WID-qEC")),
  # Sidebar layout with input and output definitions ----
  sidebarLayout(
    
    # Sidebar panel for inputs ----
    sidebarPanel(
      fileInput("file1", "Choose Results Analysis File", # Input: Select a file ----
                multiple = FALSE,
                accept = c(".txt")),
      tableOutput("resultsQuantstudio"), # for display uploaded data
      tags$hr(), # Horizontal line ----
      downloadButton("download", "Download PMR Result File") #Download button
    ),

    # Main panel for displaying part of output ----
    mainPanel(
      imageOutput("stdCurve") # display standard curve
    ))
)

# Define server logic ----
server <- function(input, output) {
  #Calculate PMR and standard curve upon uploading of files exported from QuantStudio
  myResults <- eventReactive({input$file1}, {
    # avoid displaying error message
    req(input$file1)
    
    #read in inputfiles
    data <- read.table(file = input$file1$datapath,
                        sep = "\t",
                        skip = 47,
                        header = TRUE)
    
    # ensure sample IDs are in correct format
    data$Sample.Name <- as.character(data$Sample.Name)
    
    # calculate PMR
    results <- calculate_pmr(data)
    results
  })
  
  # display uploaded data
  output$resultsQuantstudio <- renderTable({
    req(input$file1)
    data <- read.table(file = input$file1$datapath,
                       sep = "\t",
                       skip = 47,
                       header = TRUE)
    head(data)[,1:6]
  })
  
  # display calculated curve
  output$stdCurve <- renderPlot({
    myResults()[[1]]
  })
  
  # download batch results
  output$download <- downloadHandler(
    
    filename = function() {paste0("Final_results_", Sys.Date(), ".zip")},
    content = function(file){
      
      #locate temporary dir created in this R session & create sub-dir with time within
      temp_dir <- file.path(tempdir(), as.integer(Sys.time()))
      dir.create(temp_dir)
      
      # save cal curve to temporary dir
      ggsave(myResults()[[1]],
             file = paste0(temp_dir, "/","calibration_curve.png"),
             width = 5,
             height = 4) 
      
      # save CT plot to temporary dir
      ggsave(myResults()[[10]],
             file = paste0(temp_dir, "/","CTCOL2A1_MeanStdev.png"),
             width = 85,
             height = 80,
             unit = "mm") 
      
      # save batched results to temporary dir
      wb <- createWorkbook()
      
      addWorksheet(wb, "PMR Values") #PMRs
      writeDataTable(wb = wb, sheet = 1, x = myResults()[[2]], rowNames=FALSE) 
      addWorksheet(wb, "Cq mean") #mean Cq
      writeDataTable(wb = wb, sheet = 2, x = myResults()[[3]], rowNames=FALSE) 
      addWorksheet(wb, "Cq SD") #Cq stdev
      writeDataTable(wb = wb, sheet = 3, x = myResults()[[4]], rowNames=FALSE) 
      addWorksheet(wb, "Input conc") #log(copy number/5uL)
      writeDataTable(wb=wb, sheet = 4, x = myResults()[[5]], rowNames=FALSE) 
      addWorksheet(wb, "low DNA input") #samples for which COL2A1 failed in both reps, should have negative controls
      writeDataTable(wb=wb, sheet = 5, x = myResults()[[6]], rowNames=FALSE) 
      addWorksheet(wb, "Reprocessing needed") # samples for which for only one of two reps COL2A1 failed
      writeDataTable(wb = wb, sheet = 6, x = myResults()[[7]], rowNames=FALSE) 
      addWorksheet(wb, "Warning high CT values") # samples for which targets are below threshold, but still high
      writeDataTable(wb=wb, sheet = 7, x= myResults()[[8]], rowNames=FALSE)
      addWorksheet(wb, "Warning COL2A1 SD high") # samples for which SD CT COL2A1 > 1.5
      writeDataTable(wb=wb, sheet = 8, x= myResults()[[9]], rowNames=FALSE)

      saveWorkbook(wb, paste0(temp_dir, "/","batch_results.xlsx"))
      
      # save final result summary separately to temporary dir
      write.csv(myResults()[[11]], 
                file=paste0(temp_dir, "/", "final_results.csv"), 
                row.names=FALSE, 
                fileEncoding = "UTF-8")
      
      # zip up temporary dir to file
      zip::zip(
        zipfile = file,
        files = dir(temp_dir),
        root = temp_dir
      )
    },
    contentType = "application/zip"
  )
}

# Run the app ----
shinyApp(ui = ui, server = server)

src/calculate_pmr_gyn.R
# PMR calculation main work function
# contact: charlottevavourakis@gmail.com
# under development for Shiny App, Last update 11 August 2022
# this version has the correct way of calculating coefficients for multi-point calibration
# this version does not calculate the mean, until after threshold_COL2A1 checking is done
# Ct Sd is given again
# everything is calculated from the results file
# tested on the Barcelona data that the calibration correction does not make a difference in the final outcome of the WID-qEC (besides giving slightly different PMRs) compared to original pipeline published
# tested on one plate of the Hall dataset, that the fixed intercept/slope option works the same
 
# Keep option in for fixed intercept and slope, or using external calibration curve (e.g. on separate plate)
# but will have to reimplement so that in case of external calibration curve, also gblock is taken from separate plate
# 12 August 2022

calculate_pmr <- function(data,#experimentname,
                          threshold_COL2A1=35,
                          external_curve=FALSE,
                          curve=NULL,
                          fix_intercept=36.9,
                          fix_slope=-3.4,
                          calib_fixed=FALSE){
  
  # WID-qEC thresholds
  qEC_threshold1 <- 0.03
  qEC_threshold2 <- 0.63
  
  # Log message (including warning)
  #write(paste0("[pmr_calc] ", date()), file = paste0(output, experimentname, "_log.txt"), append = F)
  #write(paste0("[pmr_calc] Platform: ", .Platform$OS.type), file = paste0(output, experimentname, "_log.txt"), append = T)
  #write(msg2, file = paste0(output, experimentname, "_log.txt"), append = T)

  #if(calib_fixed == TRUE){
  #  write("\nWarning: instead of using the on-plate calibration curve, you are using FIXED values for the intercept and slope. Use this option with caution.\n",
  #              file = paste0(output, experimentname, "_log.txt"), append = T)
  #}
  
  # initialize list so multiple objects can be returned from function ----
  list_out <- list()
  
  # Check calibration with COL2A1 values ----
  # note that this section works also using variable data (instead of curve), 
  # but as written now option user has the option to upload a calibration curve from a different plate
 
  # Generate standard curve
  plot <- data %>%
    filter(Target == "COL2A1" & ((Sample %in% c("STD1", "STD2", "STD3","STD4") | 
                                    (Sample %in% c("STD_1", "STD_2", "STD_3", "STD_4")) | 
                                    (Sample %in% c("Std 1", "Std 2", "Std 3", "Std 4")) | 
                                    (Sample %in% c("Std_1", "Std_2", "Std_3", "Std_4")) | 
                                    (Sample %in% c("Std. 1", "Std. 2", "Std. 3", "Std. 4"))))) %>%
    mutate(x = case_when(Sample %in% c("STD1", "STD_1", "Std 1", "Std_1", "Std. 1") ~ 6, # concentration = log10(copy numbers/5uL)
                         Sample %in% c("STD2", "STD_2", "Std 2", "Std_2", "Std. 2") ~ 5,
                         Sample %in% c("STD3", "STD_3", "Std 3", "Std_3", "Std. 3") ~ 4,
                         Sample %in% c("STD4", "STD_4", "Std 4", "Std_4", "Std. 4") ~ 3)) %>%
    mutate(ct = ifelse(Cq == "Undetermined", NA, as.numeric(Cq)))
  
  fit <- lm(ct ~ x, data = plot)
  anno <- paste0("R2 = ", signif(summary(fit)$r.squared, 3),
                 "\ny = ", round(fit$coefficients[2],2) , "x + ", round(fit$coefficients[1],2))
  y = max(plot$ct) - 0.05*max(plot$ct)
  intercept <- unname(fit$coefficients[1])
  slope <- unname(fit$coefficients[2])
  R2 <- summary(fit)$r.squared
  
  # Plot calibration curve
  curveplot <- plot %>%
    ggplot(aes(x = x,
               y = ct)) +
    geom_smooth(method = "lm",
                formula = y ~ x,
                size = 0.5,
                se = FALSE,
                alpha = 0.3,
                colour = "gray40") +
    geom_point(aes(colour = Sample),
               size = 0.75) +
    theme(legend.title = element_blank(),
          legend.position = "top",
          legend.key = element_blank()) +
    xlab("Concentration \n [ log(copy number/5uL) ]") +
    ylab("Ct value")  +
    annotate("text",
             x = 5,
             y = 28,
             label = anno,
             hjust = 0,
             size = 3,
             colour = ifelse(R2 < 0.99, "red", "black"))
  
  # add plot to list
  list_out[[1]] <- curveplot
  
  rm(plot, curveplot)
  
  # If user choose fixed values overwrite the coefficients from standard curve
  if(calib_fixed==TRUE){
    intercept <- fix_intercept
    slope <- fix_slope
  }
  
  # reformat data in results file based on thresholds and replication ----
  
  # Carry on with Cq means
  data <- data %>%
    mutate(rep = case_when(grepl("A|C|E|G|I|K|M|O", Well.Position) == TRUE ~ 1,
                           grepl("B|D|F|H|J|L|N|P", Well.Position) == TRUE ~ 2)) %>%
    mutate(ct = as.numeric(ifelse(Cq == "Undetermined", NA, Cq)))
  
  # collect target and sample overviews
  targets <- unique(data$Target)
  targets2 <- targets[!targets %in% "COL2A1"]
  samples <- unique(data$Sample)
  
  # For each rep evaluate COL2A1 and add a column PASS/FAIL for filtering data
  df <- NULL
  for (s in samples) {
    subset1 <- data %>%
      filter(Sample==s,
             rep==1,
             Target=="COL2A1")
    subset1 <- subset1 %>%
      mutate(COL2A1_check =
               case_when( 
      is.na(ct) | ct > threshold_COL2A1 ~ "FAIL", TRUE ~ "PASS")) %>%
      dplyr::select(Sample,rep,COL2A1_check)
    df <- rbind(df,subset1)
    
    subset2 <- data %>%
      filter(Sample==s,
             rep==2,
             Target=="COL2A1")
    subset2 <- subset2 %>%
      mutate(COL2A1_check = 
               case_when( 
      is.na(ct) | ct > threshold_COL2A1 ~ "FAIL", TRUE ~ "PASS")) %>%
      dplyr::select(Sample,rep,COL2A1_check)
    df <- rbind(df,subset2)
  }
  
  #if(unique(df[df$Sample=="H2O",]$COL2A1_check) != "FAIL"){
  #  write("WARNING: COL2A1 amplified in the negative control.", file = paste0(output, experimentname, "_log.txt"), append = T)
  #}
  
  data <- left_join(data, df, by=c("Sample","rep"))
  
  # When both reps of a sample are undetermined or above the set threshold for COL2A1,
  # DNA input for this sample was too low, and no PMR could be calculated.
  # If one out of two reps is undetermined or above the set threshold,
  # sample should to be reprocessed (COL2A1) or reprocessing is recommended. 
  # In these cases, one Ct value is taken forward. 
  low_input_fail <- NULL # negative control should always endup here
  reprocess_needed <- NULL
  reprocess <- NULL
  
  for (s in samples){
    for(t in targets){
      tmp <- data %>%
        filter(Sample == s & Target == t) %>%
        dplyr::select(Sample, rep, ct, COL2A1_check)
      
      if(t == "COL2A1" & all(tmp$COL2A1_check=="FAIL")){
        # if target is COL2A1 and both reps fail, sample should be listed under low_input_fail
        x <- data.frame(sample = tmp$Sample, COL2A1_Ct = tmp$ct)
        low_input_fail <- rbind(low_input_fail, x)
      }
      
      if(any(is.na(tmp$ct)) && any(!is.na(tmp$ct))){
        # Carry one Ct value forward
        und <- sum(is.na(tmp$ct))
        ind1 <- na.omit(data$Sample == s & data$Target == t)
        ct <- ifelse(und < 2, na.omit(tmp$ct), NA)
        data[ind1,]$ct <- ct
        
        if(t == "COL2A1"){
          # reprocessing required
          x <- data.frame(sample = tmp$Sample[is.na(tmp$ct)],
                          rep = tmp$rep[is.na(tmp$ct)],
                          target = t,
                          `COL2A1_check (this replicate)` = tmp$COL2A1_check[is.na(tmp$ct)])
          reprocess_needed <- rbind(reprocess_needed,x)
        } else {
          # reprocessing recommended
          x <- data.frame(sample = tmp$Sample[is.na(tmp$ct)],
                          rep = tmp$rep[is.na(tmp$ct)],
                          target = t,
                          `COL2A1_check (this replicate)` = tmp$COL2A1_check[is.na(tmp$ct)])
          reprocess <- rbind(reprocess,x)
        }
      }
    }
  }
  
  
  # Log messages
  #write(ifelse(sum(low_input_fail$sample=="H2O")==2,
  #             paste0("Samples with insufficient input DNA (plus 2 H2O controls): ", nrow(low_input_fail)-2),
  #             paste0("Not all H2O controls failed to amplify COL2A1. Reprocessing strongly recommended.")),
  #      file = paste0(output, experimentname, "_log.txt"), append = T)
  
  #if(!is_empty(reprocess_needed)){
  #  write(paste0("Samples that failed to amplify COL2A1 in one replicate (reprocessing needed): ", nrow(reprocess_needed)),
  #             file = paste0(output, experimentname, "_log.txt"), append = T)
  #}
  
  #if(!is_empty(reprocess)){
  #  write(paste0("Samples that failed to amplify one of the targets in one replicate (reprocessing recommended): ", nrow(reprocess)),
  #        file = paste0(output, experimentname, "_log.txt"), append = T)
  #}
  
  data1 <- data %>%
    pivot_wider(id_cols = Sample,
                names_from = Target,
                values_from = ct,
                values_fn = mean)
  
  data2 <- data %>%
    pivot_wider(id_cols = Sample,
                names_from = Target,
                values_from = ct,
                values_fn = sd)

  # Calculate PMR ----
  
  # Calculate input amounts (using calibration curve COL2A1)
  for (t in targets){
    varname <- paste0("input_",t)
    data1[varname] <- with(data1, 10^((data1[t]-intercept)/slope))
  }
  
  # Normalization: Divide calculated input amount by input amount calculated for COL2A1
  for (t in targets){
    varname <- paste0("ref_",t)
    calcname <- paste0("input_",t)
    data1[varname] <- data1[calcname]/data1["input_COL2A1"]
  }
  
  # Fully methylated DNA (SSS1 or gBlock) - only gBlock implemented
  gblock <- data1 %>%
    filter(Sample %in% c("gBlock", "gBLOCK", "gBlock (+)")) 
  
  # For each ref gene divide normalized input sample by normalized input fully methylated DNA *100
  results <- data1
  
  for (t in targets2){
    varname <- t
    calcname <- paste0("ref_",t)
    div_block <- c(gblock[calcname])
    results[varname] <- (data1[calcname]/div_block)*100
  }
  
  results <- results %>%
    select(Sample, all_of(targets2))
  results[is.na(results)] <- 0 # set all NAs to 0
  
  # Calculate concentration input DNA based on COL2A mean Ct
  conc_input <- data1 %>% 
    as.data.frame() %>%
    mutate(input_conc = (data1$COL2A1 - intercept)/slope) %>%
    select(Sample, input_conc) %>% droplevels
  
  # Calculate WID-qEC and or WID-CIN
  targets_qEC <- c("GYPC1","GYPC2","ZSCAN12")
  if(all(targets_qEC %in% targets)){
    results <- results %>%
    mutate(WIDqEC = (GYPC1 + GYPC2 + ZSCAN12)) # calculate sumPMR for WID-qEC
    results <- results %>% 
      mutate(WIDqEC_test = case_when(results$WIDqEC < qEC_threshold1 ~ "low risk EC/CIN",
                                     results$WIDqEC >= qEC_threshold1 & results$WIDqEC <= qEC_threshold2 ~ "high risk EC/CIN",
                                     results$WIDqEC > qEC_threshold2 ~ "very high risk EC/CIN"))
  }
  
  targets_qCIN <- c("RALYL", "DPP6", "GSX1")
  if(all(targets_qCIN %in% targets)){
    results <- results %>%
      mutate(WIDqCIN = (RALYL + DPP6 + GSX1)) # calculate sumPMR for WID-qCIN
  }
  
  # Save results to list ---
  results <- as.data.frame(results)
  
  data1 <- data1 %>%
    select(Sample, COL2A1, all_of(targets2)) %>%
    as.data.frame()
  
  data2 <- data2 %>%
    select(Sample, COL2A1, all_of(targets2)) %>%
    as.data.frame()
  
  list_out[[2]] <- results #PMR + optional WIDqEC + WIDqEC outcome +WIDqCIN
  list_out[[3]] <- data1 #mean Cq
  list_out[[4]] <- data2 #stdev Cq
  list_out[[5]] <- conc_input #log(copy number/5uL)
  list_out[[6]] <- low_input_fail #samples for which COL2A1 failed in both reps, should have negative controls

  if(!is_empty(reprocess_needed)){ # samples for which for only one of two reps COL2A1 failed
    list_out[[7]] <- reprocess_needed
  } else {
    list_out[[7]] <- as.data.frame("none")
  }
  
  if(!is_empty(reprocess)){ # samples for which for only one of two reps target amplified
    list_out[[8]] <- reprocess
  } else {
    list_out[[8]] <- as.data.frame("none")
  }
  
  return(list_out)
}


src/calculate_pmr_widqec.R
# PMR calculation main work function
# contact: charlottevavourakis@gmail.com
# under development for Shiny App version for TirolPath
# Last update 19 January 2023

calculate_pmr <- function(data,#experimentname,
                          threshold_COL2A1=30,
                          threshold_targets=35,
                          threshold_targets_warning=30,
                          external_curve=FALSE, #these and parameters below are not used
                          curve=NULL,
                          fix_intercept=36.9,
                          fix_slope=-3.4,
                          calib_fixed=FALSE){
  
  # WID-qEC thresholds
  qEC_threshold1 <- 0.03
  #qEC_threshold2 <- 0.63
  
  # initialize list so multiple objects can be returned from function ----
  list_out <- list()
  
  # Check calibration with COL2A1 values ----
  
  # Generate standard curve
  plot <- data %>%
    filter(Target.Name == "COL2A1" & ((Sample.Name %in% c("STD1", "STD2", "STD3","STD4") | 
                                    (Sample.Name %in% c("STD_1", "STD_2", "STD_3", "STD_4")) | 
                                    (Sample.Name %in% c("Std 1", "Std 2", "Std 3", "Std 4")) | 
                                    (Sample.Name %in% c("Std_1", "Std_2", "Std_3", "Std_4")) | 
                                    (Sample.Name %in% c("Std. 1", "Std. 2", "Std. 3", "Std. 4"))))) %>%
    mutate(x = case_when(Sample.Name %in% c("STD1", "STD_1", "Std 1", "Std_1", "Std. 1") ~ 6, # concentration = log10(copy numbers/5uL)
                         Sample.Name %in% c("STD2", "STD_2", "Std 2", "Std_2", "Std. 2") ~ 5,
                         Sample.Name %in% c("STD3", "STD_3", "Std 3", "Std_3", "Std. 3") ~ 4,
                         Sample.Name %in% c("STD4", "STD_4", "Std 4", "Std_4", "Std. 4") ~ 3)) %>%
    mutate(ct = ifelse(CT == "Undetermined", NA, as.numeric(CT)))
  
  fit <- lm(ct ~ x, data = plot)
  anno <- paste0("R2 = ", signif(summary(fit)$r.squared, 3),
                 "\ny = ", round(fit$coefficients[2],2) , "x + ", round(fit$coefficients[1],2))
  y = max(plot$ct) - 0.05*max(plot$ct)
  intercept <- unname(fit$coefficients[1])
  slope <- unname(fit$coefficients[2])
  R2 <- summary(fit)$r.squared
  
  # Plot calibration curve
  curveplot <- plot %>%
    ggplot(aes(x = x,
               y = ct)) +
    geom_smooth(method = "lm",
                formula = y ~ x,
                size = 0.5,
                se = FALSE,
                alpha = 0.3,
                colour = "gray40") +
    geom_point(aes(colour = Sample.Name),
               size = 0.75) +
    theme(legend.title = element_blank(),
          legend.position = "top",
          legend.key = element_blank()) +
    xlab("Concentration \n [ log(copy number/5uL) ]") +
    ylab("Ct value")  +
    annotate("text",
             x = 5,
             y = 28,
             label = anno,
             hjust = 0,
             size = 3,
             colour = ifelse(R2 < 0.99, "red", "black"))
  
  # add plot to list
  list_out[[1]] <- curveplot
  
  rm(plot, curveplot)
  
  # If user choose fixed values overwrite the coefficients from standard curve
  if(calib_fixed==TRUE){
    intercept <- fix_intercept
    slope <- fix_slope
  }
  
  # reformat data in results file based on thresholds and replication ----
  data <- data %>%
    mutate(rep = case_when(grepl("A|C|E|G|I|K|M|O", Well.Position) == TRUE ~ 1,
                           grepl("B|D|F|H|J|L|N|P", Well.Position) == TRUE ~ 2)) %>%
    mutate(ct = as.numeric(ifelse(CT > threshold_targets | Amp.Status=="No Amp", NA, CT))) # set suspicious amplifications to NA for targets
  
  # set suspicious amplifications to NA for COL2A1
  ind <- which(data$Target.Name == "COL2A1" & !data$Sample.Name %in% c("NTC_H2O","NTC", "H2O") & (data$CT > threshold_COL2A1 | data$Amp.Status=="No Amp"))
  if(!is_empty(ind)){
    data[ind,]$ct <- NA
  }
  
  # collect target and sample overviews 
  targets <- unique(data$Target.Name)
  targets2 <- targets[!targets %in% "COL2A1"]
  samples <- unique(data$Sample.Name)
  samplesNoctrl <- setdiff(samples,c("STD1", "STD2", "STD3","STD4","STD_1", "STD_2", "STD_3", "STD_4","Std 1", "Std 2", "Std 3", "Std 4","Std_1", "Std_2", "Std_3", "Std_4","Std. 1", "Std. 2", "Std. 3", "Std. 4","NTC_H2O","NTC", "H2O")) #but include gBlocks C("posCo","PosCo","gBlock", "gBLOCK", "gBlock (+)")
  
  # For each rep evaluate COL2A1 and add a column PASS/FAIL for filtering data
  df <- NULL
  for (s in samples) {
    subset1 <- data %>%
      filter(Sample.Name==s,
             rep==1,
             Target.Name=="COL2A1")
    subset1 <- subset1 %>%
      mutate(COL2A1_check =
               case_when( 
      is.na(ct) ~ "FAIL", TRUE ~"PASS")) %>%
      #is.na(ct) | ct > threshold_COL2A1 ~ "FAIL", TRUE ~ "PASS")) %>%
      dplyr::select(Sample.Name,rep,COL2A1_check)
    df <- rbind(df,subset1)
    
    subset2 <- data %>%
      filter(Sample.Name==s,
             rep==2,
             Target.Name=="COL2A1")
    subset2 <- subset2 %>%
      mutate(COL2A1_check = 
               case_when( 
      is.na(ct) ~ "FAIL", TRUE ~"PASS")) %>%
      #is.na(ct) | ct > threshold_COL2A1 ~ "FAIL", TRUE ~ "PASS")) %>%
      dplyr::select(Sample.Name,rep,COL2A1_check)
    df <- rbind(df,subset2)
  }
  
  data <- left_join(data, df, by=c("Sample.Name","rep"))
  
  # When both reps of a sample are undetermined or above the set threshold for COL2A1,
  # DNA input for this sample was too low, and no PMR could be calculated.
  # If one out of two reps is undetermined or above the set threshold for COL2A1, the sample should be reprocessed.
  # IF one out of two reps is undetermined or above the set threshold for the other targets, the respective amplification is disregarded (no ct value is taken forward) 
  low_input_fail <- NULL # negative control should always endup here
  reprocess_needed <- NULL # one of two COL2A1 reps failed
  
  for (s in samples){
    for(t in targets){
      tmp <- data %>%
        filter(Sample.Name == s & Target.Name == t) %>%
        dplyr::select(Sample.Name, rep, ct, COL2A1_check)
      
      if(any(is.na(tmp$ct)) && any(!is.na(tmp$ct))){
        # set non-replicated amplification to NA
        ind <- which((data$Sample.Name == s & data$Target.Name == t & !is.na(data$ct)))
        data[ind,]$ct <- NA
        
        if(t == "COL2A1"){
          # list sample as reprocessing required
          x <- data.frame(sample = tmp$Sample.Name[is.na(tmp$ct)],
                          rep = tmp$rep[is.na(tmp$ct)],
                          target = t,
                          `COL2A1_check (this replicate)` = tmp$COL2A1_check[is.na(tmp$ct)])
          reprocess_needed <- rbind(reprocess_needed,x)
        }}
      
      if(t == "COL2A1" & all(tmp$COL2A1_check=="FAIL")){
        # if target is COL2A1 and both reps fail, sample should be listed under low_input_fail
        x <- data.frame(sample = tmp$Sample.Name, COL2A1_Ct = tmp$ct)
        low_input_fail <- rbind(low_input_fail, x)
        }
    }
  }
  
  # Flag reactions with high ct (after removing spurious amplification)
  flagged <- data %>% 
    mutate(flag=case_when(ct >= threshold_targets_warning ~ paste0("CT below threshold ", threshold_targets,", but still higher than ",threshold_targets_warning), TRUE ~"")) %>%
    select(Well.Position,Sample.Name,Target.Name,rep,ct,COL2A1_check,flag) %>%
    filter(flag != "")
  
  # Carry on with ct means
  data1 <- data %>%
    pivot_wider(id_cols = Sample.Name,
                names_from = Target.Name,
                values_from = ct,
                values_fn = mean)
  
  data2 <- data %>%
    pivot_wider(id_cols = Sample.Name,
                names_from = Target.Name,
                values_from = ct,
                values_fn = sd)

  # Calculate PMR ----
  
  # Calculate input amounts (using calibration curve COL2A1)
  for (t in targets){
    varname <- paste0("input_",t)
    data1[varname] <- with(data1, 10^((data1[t]-intercept)/slope))
  }
  
  # Normalization: Divide calculated input amount by input amount calculated for COL2A1
  for (t in targets){
    varname <- paste0("ref_",t)
    calcname <- paste0("input_",t)
    data1[varname] <- data1[calcname]/data1["input_COL2A1"]
  }
  
  # Fully methylated DNA (SSS1 or gBlock) - only gBlock implemented
  gblock <- data1 %>%
    filter(Sample.Name %in% c("posCo","PosCo","gBlock", "gBLOCK", "gBlock (+)")) 
  
  # For each ref gene divide normalized input sample by normalized input fully methylated DNA *100
  results <- data1
  
  for (t in targets2){
    varname <- t
    calcname <- paste0("ref_",t)
    div_block <- c(gblock[calcname])
    results[varname] <- (data1[calcname]/div_block)*100
  }
  
  results <- as.data.frame(results)
  results <- results %>%
    select(Sample.Name, all_of(targets2))
  results[is.na(results)] <- 0 # set all NAs to 0
  
  # Calculate concentration input DNA based on COL2A mean Ct
  conc_input <- data1 %>% 
    as.data.frame() %>%
    mutate(input_conc = (data1$COL2A1 - intercept)/slope) %>%
    select(Sample.Name, input_conc) %>% droplevels
  
  # Calculate WID-qEC
  targets_qEC <- c("GYPC1","GYPC2","ZSCAN12")
  if(all(targets_qEC %in% targets)){
    results <- results %>%
    mutate(WIDqEC = round((GYPC1 + GYPC2 + ZSCAN12), digits=3)) # calculate sumPMR for WID-qEC
    results <- results %>% 
      mutate(WIDqEC_interpret = case_when(results$WIDqEC >= qEC_threshold1 ~ "Positiv",
                                     TRUE ~ "Negativ"))
  }
  
  # samples for which both reps COL2A1 failed, PMR should not be 0, but should be NA
  if(!is_empty(low_input_fail)){
    samples_failed <- low_input_fail %>% filter(!sample %in% c("NTC","NTC_H2O")) %>% unique() %>% pull(sample)
    if(!is_empty(samples_failed)){
      for (i in 2:(length(targets2)+3)){
        for (j in 1:length(samples_failed)){
          results[results$Sample.Name==samples_failed[j],i]<- NA
        }
      }
    }
  }
  
  # samples for which one rep of COL2A1 failed, PMR should not be 0, but should be NA
  if(!is_empty(reprocess_needed)){
    samples_failed <- reprocess_needed %>% pull(sample)
    for (i in 2:(length(targets2)+3)){
      for (j in 1:length(samples_failed)){
        results[results$Sample.Name==samples_failed[j],i]<- NA
      }
    }
  }
    
  # collect results mean+SD CT
  data1 <- data1 %>%
    select(Sample.Name, COL2A1, all_of(targets2)) %>%
    as.data.frame()
  
  data2 <- data2 %>%
    select(Sample.Name, COL2A1, all_of(targets2)) %>%
    as.data.frame()
  
  # Additional warning for When COL2A1 passed for both reps but CT stdev > 1.5 CT
  warning = data2 %>% 
    select(Sample.Name,COL2A1) %>%
    mutate(Warning.SD.COL2A1 = case_when(COL2A1 > 1.5 ~ "warning SD > 1.5 CT", TRUE ~ "PASS"))
  
  if(any(warning$Warning.SD.COL2A1== "warning SD > 1.5 CT")){
    warning = warning %>% filter(Warning.SD.COL2A1=="warning SD > 1.5 CT")
  } else{
    warning = NULL
  }
  
  #plot
  ### transform data frames to long format, combine dataframes, drop NAs, select samples only
  Ct_means <- pivot_longer(data1,!Sample.Name, names_to = "Target", values_to = "Ct_mean")
  Ct_sd <- pivot_longer(data2,!Sample.Name, names_to = "Target", values_to = "Ct_sd")
  df <- full_join(Ct_means, Ct_sd) %>%
    drop_na() %>%
    filter(Sample.Name %in% samplesNoctrl)
  
  ### plot output depends on number of 
  # plot processed data all targets and save
  plot <- df %>%
    filter(Target=="COL2A1") %>%
    ggplot(aes(x=Sample.Name, y=Ct_mean)) +
    geom_point()+
    geom_errorbar(aes(ymin=Ct_mean-Ct_sd, ymax=Ct_mean+Ct_sd), width=0.5) +
    #facet_wrap(~target) +
    xlab("") + #rename axis text label
    ylab("Ct (mean)") +
    theme_minimal() +
    theme(axis.title.x=element_text(size=12),
          axis.title.y=element_text(size=12),
          axis.text.x=element_text(size=5,angle=60,hjust=1),
          axis.text.y=element_text(size=12),
          legend.text=element_text(size=12),
          legend.title = element_blank(),
          legend.key.size= unit(6, "mm"))
  
  # make the final summary
  final <- results %>%
    filter(!Sample.Name %in% c("STD_1","STD_2","STD_3","STD_4","posCo","PosCo","NTC_H2O","NTC")) %>%
    select(Sample.Name, WIDqEC, WIDqEC_interpret) %>%
    dplyr::rename(SampleName = Sample.Name)
  final$WIDqEC <- format(final$WIDqEC, nsmall=3, scientific=FALSE)
  
  samples <- samples[!samples %in% c("STD_1","STD_2","STD_3","STD_4","posCo","PosCo","NTC_H2O","NTC")]
  
  QC <- data.frame(SampleName = samples)
  
  QC <- QC %>%
    mutate(QC=case_when(
     SampleName %in% low_input_fail$sample ~ "Insufficient DNA in both reps", #COL2A1 did not amplify in any of the reps 
     SampleName %in% reprocess_needed[,1] ~ "Reprocessing needed insufficient DNA in one rep", # samples for which for only one of two reps COL2A1 failed
     SampleName %in% warning[,1] ~ "STDEV CT COL2A1 exceeds 1.5 cycles",
     TRUE ~ "PASS" #all ok
    ))
  
  final <- full_join(final,QC)
  
  # Save results to list ---
  list_out[[2]] <- results #PMR + WIDqEC + WIDqEC outcome
  list_out[[3]] <- data1 #mean CT
  list_out[[4]] <- data2 #stdev CT
  list_out[[5]] <- conc_input #log(copy number/5uL)
  if(is_empty(low_input_fail)){
    list_out[[6]] <- as.data.frame("you have some amplification in the NTC_H2O")
  } else{
    list_out[[6]] <- low_input_fail #samples for which COL2A1 failed in both reps, should have negative controls
  }
  
  if(!is_empty(reprocess_needed)){ # samples for which one of two reps COL2A1 failed
    list_out[[7]] <- reprocess_needed
  } else {
    list_out[[7]] <- as.data.frame("none")
  }
  
  if(!is_empty(flagged)){ #reactions with high CT, current threshold set at 30
    list_out[[8]] = flagged
  } else {
    list_out[[8]] = as.data.frame("none")
  }
  
  if(!is_empty(warning)){ #marks samples for which STDEV COL2A > 1.5 CT (QC warning)
    list_out[[9]] = warning
  } else {
    list_out[[9]] = as.data.frame("none")
  }
  
  list_out[[10]] <- plot #mean CT COL2A1
  
  list_out[[11]] <- final #information in final csv file
  
  return(list_out)
}
